### PR TITLE
feat(sink): support async for bigquery sink 

### PR DESCRIPTION
**Pull request by xxhZs**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

in https://github.com/risingwavelabs/risingwave/issues/17383#issuecomment-2190969273
We found that waiting to write data can block for a long time, affecting throughput.
So we support async for bigquery sink

bench with this pr:
avg: 108025 rows/s 
p90: 116736 rows/s 
p95: 116736 rows/s 
p99: 118784 rows/s

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/connector/src/sink/big_query.rs [hunk 1]
```diff
@@ -582,80 +565,64 @@ impl BigQuerySinkWriter {
...
-
-    async fn update_vnode_bitmap(&mut self, _vnode_bitmap: Arc<Bitmap>) -> Result<()> {
+        let rows = AppendRowsRequestRows::ProtoRows(ProtoData {
+            writer_schema: Some(self.writer_pb_schema.clone()),
+            rows: Some(ProtoRows { serialized_rows }),
+        });
+        let (expect_offset, mut rx) = self.client.get_subscribe();
+        let future = Box::pin(async move {
+            loop {
+                match rx.recv().await {
```
Comment (wenym1) [inline src/connector/src/sink/big_query.rs:595]: I think instead you can create a oneshot channel for each AppendRowsRequest  so that you won't need this shared broadcast channel. When the spawned worker knows that a request is handled, it notifies the rx with the oneshot tx, and here the future can be simply the rx.

File: src/connector/src/sink/big_query.rs [hunk 2]
```diff
@@ -678,49 +645,74 @@ impl StorageWriterClient {
...
-            .append_rows(Request::new(tokio_stream::iter(append_req)))
-            .await
-            .map_err(|e| SinkError::BigQuery(e.into()))?
-            .into_inner();
-        while let Some(append_rows_response) = resp
-            .message()
+        let append_req = AppendRowsRequest {
+            write_stream: write_stream.clone(),
+            offset: None,
+            trace_id: Uuid::new_v4().hyphenated().to_string(),
```
Comment (wenym1) [inline src/connector/src/sink/big_query.rs:708]: Can you tracked the `trace_id` locally instead of incrementing the `self.offset`? I think in the response stream, it should return the same trace_id in the corresponding response.

**Review by wenym1 — COMMENTED** [review]
Actually we can avoid spawning a tokio task to poll the response stream. We can implement our own `BigQueryLogSinker`, which polls the log_reader and the response_stream with a `select`.

File: src/connector/src/sink/big_query.rs (hunk 2 above)
Comment (xxhZs) [reply src/connector/src/sink/big_query.rs:778]: Tried it, but there is no ID in the returned message

File: src/connector/src/sink/big_query.rs [hunk 3]
```diff
@@ -82,23 +87,116 @@ pub struct BigQueryCommon {
...
+        }
+        Ok(())
+    }
+
+    pub async fn wait_next_offset(&mut self) -> Result<TruncateOffset> {
+        loop {
+            if let Some(offset) = self.wait_one_offset().await? {
+                return Ok(offset);
+            } else {
+                self.rx.recv().await.unwrap();
```
Comment (wenym1) [inline src/connector/src/sink/big_query.rs:132]: The `tx` and `rx` is not needed. When we await on `self.rx.recv()`, we have the ownership of `self`, and no one will be able to access `self.tx` to notify anything.

The future manager can be very simple. It only needs two methods
```
fn add_offset(&mut self, offset: TruncateOffset) {
    self.offset_queue.push_back(offset);
}

fn wait_next_offset(&mut self) -> Result<TruncateOffset> {
    if let Some(TruncateOffset::Barrier(_)) = self.offset_queue.front() {
        return Ok(self.offset_queue.pop_front().unwrap());
    }
    let resp = self.resp_stream.try_next().await?.ok_or_else("end of stream")?;
    self.offset_queue.pop_front().ok_or_else("should have pending chunk offset when we receive new response")
}
```

We can limit the number of pending request by not polling the log reader in the `select!` by
```
select!(
    self.future_manager.wait_next_offset() => { ... }
    self.log_reader.next_item() if self.future_manager.offset_queue.len() <= MAX_XXX => { ... }
)
```

**Comment by xxhZs** [comment]
Since each row of bg can't exceed 10MB, new logic is added to split the size of each row

File: src/connector/src/sink/big_query.rs [hunk 4]
```diff
@@ -82,23 +86,97 @@ pub struct BigQueryCommon {
...
+            resp_stream: Box::pin(resp_stream),
+        }
+    }
+
+    pub fn add_offset(&mut self, offset: TruncateOffset, mut resp_num: usize) {
+        while resp_num > 1 {
+            self.offset_queue.push_back(None);
+            resp_num -= 1;
+        }
+        self.offset_queue.push_back(Some(offset));
```
Comment (stdrc) [inline src/connector/src/sink/big_query.rs:115]: Why inserting `resp_num - 1` `None`s and 1 `Some(offset)`? Could you please add some comments here?

File: src/connector/src/sink/big_query.rs [hunk 5]
```diff
@@ -82,23 +86,97 @@ pub struct BigQueryCommon {
...
+    }
 
-fn default_retry_times() -> usize {
-    5
+    pub async fn wait_next_offset(&mut self) -> Result<Option<TruncateOffset>> {
```
Comment (stdrc) [inline src/connector/src/sink/big_query.rs:118]: I guess `async fn next_offset` is enough to indicate that the function call may need to be "waited".

File: src/connector/src/sink/big_query.rs [hunk 6]
```diff
@@ -82,23 +86,97 @@ pub struct BigQueryCommon {
...
+        }
+        self.resp_stream
+            .next()
+            .await
+            .ok_or_else(|| SinkError::BigQuery(anyhow::anyhow!("end of stream")))??;
+        self.offset_queue.pop_front().ok_or_else(|| {
+            SinkError::BigQuery(anyhow::anyhow!(
+                "should have pending chunk offset when we receive new response"
+            ))
+        })
```
Comment (stdrc) [inline src/connector/src/sink/big_query.rs:130]: Why we can only directly pop and return if the front of queue is `TruncateOffset::Barrier`? Is it possible to have other variants of `TruncateOffset` in the queue? And what happen if there are?

Could you please add some comments here?

File: src/connector/src/sink/big_query.rs [hunk 7]
```diff
@@ -82,23 +86,97 @@ pub struct BigQueryCommon {
...
+    }
+}
+pub struct BigQueryLogSinker {
+    writer: BigQuerySinkWriter,
+    bigquery_future_manager: BigQueryFutureManager,
```
Comment (stdrc) [inline src/connector/src/sink/big_query.rs:135]: Why not `future_manager`? we don't need to add so many prefixes in private fields/types. And I guess it may not be so suitable to call it "manager" here.

Also, do we really need a `BigQueryFutureManager` type here? It seems don't "manage" anything, we still directly access `bigquery_future_manager.offset_queue` below😂

File: src/connector/src/sink/big_query.rs (hunk 4 above)
Comment (xxhZs) [reply src/connector/src/sink/big_query.rs:115]: Because if a chunk is too large, we will split it into multiple async tasks, so we need to wait until all the tasks are finished before we can truncate, here none means that he is only the middle async task is completed, to the chunk before you can truncate.
And i will add comments.

File: src/connector/src/sink/big_query.rs (hunk 6 above)
Comment (xxhZs) [reply src/connector/src/sink/big_query.rs:130]: Since we don't go to the bg sink to create an async task when we receive a barrier, there's no need to wait here for the
And i will add comments.

File: src/connector/src/sink/big_query.rs (hunk 7 above)
Comment (xxhZs) [reply src/connector/src/sink/big_query.rs:135]: For async sink, we have a FutureManager to manage the asyn tasks and truncate our log store chunk after the async tasks are done. this BigQueryFutureManager  has the same tasks as FutureManager, just customized for bg sink

File: src/connector/src/sink/big_query.rs (hunk 6 above)
Comment (stdrc) [reply src/connector/src/sink/big_query.rs:136]: > Since we don't go to the bg sink to create an async task when we receive a barrier, there's no need to wait here for the

Can't quite get it. Could you elaborate with proper English?

Comment (xxhZs) [reply src/connector/src/sink/big_query.rs:136]: There 3 cases:
1. Some(barrier), since we won't be sending a barrier to bigquery, there is no need to wait for the. But we will record the barriers in the log store, so we need to truncate.
2. Some(chunk), since we send a chunk to bigquery and record it in the log store, so we need to truncate and wait for resp.
3. None, It means that we split a large chunk of RW into multiple requests for bigquery (for this chunk, in offset_queue, we set it to none , none ... chunk). So we need to wait for all the none and chunk resps and truncate in Some(chunk)

File: src/connector/src/sink/big_query.rs [hunk 8]
```diff
@@ -678,49 +773,39 @@ impl StorageWriterClient {
...
         .await
         .map_err(|e| SinkError::BigQuery(e.into()))?;
-        let client = conn.conn();
-        Ok(StorageWriterClient {
-            client,
-            environment,
-        })
+        let mut client = conn.conn();
+
+        let (tx, rx) = mpsc::channel(BIGQUERY_SEND_FUTURE_BUFFER_MAX_SIZE);
```
Comment (wenym1) [inline src/connector/src/sink/big_query.rs:778]: We can use bounded channel here since we already limit the queue size in the `select!`

File: src/connector/src/sink/big_query.rs [hunk 9]
```diff
@@ -583,82 +669,91 @@ impl BigQuerySinkWriter {
...
+        >
+        + 'static
+        + Send,
+) {
+    let mut resp_stream = resp_stream
+        .await
+        .map_err(|e| SinkError::BigQuery(e.into()))?
+        .into_inner();
+    loop {
+        if let Some(append_rows_response) = resp_stream
```
Comment (wenym1) [inline src/connector/src/sink/big_query.rs:726]: When we reach the end of the `resp_stream`, we will keep yielding (). This is not correct. Instead of using `if let Some(...)`, we'd better just `resp_stream.message().await.ok_or_else(|| ...)` to turn `None` into an end-of-stream error.

File: src/connector/src/sink/big_query.rs [hunk 10]
```diff
@@ -82,23 +86,103 @@ pub struct BigQueryCommon {
...
+        self.offset_queue.push_back(Some(offset));
+    }
 
-fn default_retry_times() -> usize {
-    5
+    pub async fn next_offset(&mut self) -> Result<Option<TruncateOffset>> {
+        if let Some(Some(TruncateOffset::Barrier { .. })) = self.offset_queue.front() {
+            return Ok(self.offset_queue.pop_front().unwrap());
+        }
+        self.resp_stream
```
Comment (wenym1) [inline src/connector/src/sink/big_query.rs:128]: Here the correctness depends on that the number of non-barrier items (either None or Some(TruncateOffset::Chunk)) in the queue is the same as the inflight request. However, in the implementation, it's possible to have 0 `resp_num` when we write a chunk. For this chunk, there is no inflight request, but it will have an item in the queue, which causes inconsistency between the number of queue items and inflight requests.

I think instead of using none to represent that a chunk is split into multiple requests, we'd better store (TruncateOffset, remaining_resp_num) as the queue item. Code will be like the following
```
if let Some((offset, remaining_resp_num)) = self.offset_queue.front_mut() {
    if *remaining_resp_num == 0 {
        return Ok(self.offset_queue.pop_front().unwrap().0);
    }
    while *remaining_resp_num > 0 {
        self.resp_stream.next().await...??;
        *remaining_resp_num -= 1;
    }
} else {
    return pending().await;
}
```

File: src/connector/src/sink/big_query.rs (hunk 9 above)
Comment (xxhZs) [reply src/connector/src/sink/big_query.rs:726]: It doesn't return () at the end of the stream, after receiving every reply, if there are no errors, it returns a meaningless (), which calls the method of the stream, which reaches the end of the stream.

Comment (wenym1) [reply src/connector/src/sink/big_query.rs:726]: I know the purpose of returning a meaningless `()`. What I meant was that, when `resp_stream.message()` returns None, we should break the loop with an `EndOfStream` error, or just simply break the loop, instead of ignoring it. 

In current code, after resp_stream returns None, an `()` will still be yielded, and later when polled again, and resp_stream returns None again, and an `()` is still yielded, and this will be repeated endlessly over and over again, while the external code is not aware that the response stream has actually stopped.

File: src/bench/sink_bench/sink_option.yml [hunk 11]
```diff
@@ -100,8 +100,8 @@ Starrocks:
 BigQuery:
```
Comment (wenym1) [inline src/bench/sink_bench/sink_option.yml:100]: Revert the change.

File: src/connector/src/sink/big_query.rs [hunk 12]
```diff
@@ -82,23 +87,100 @@ pub struct BigQueryCommon {
...
-    pub retry_times: usize,
     #[serde(default)] // default false
     #[serde_as(as = "DisplayFromStr")]
     pub auto_create: bool,
 }
 
-fn default_max_batch_rows() -> usize {
-    1024
+struct BigQueryFutureManager {
+    // `offset_queue` holds the Some corresponding to each future.
```
Comment (wenym1) [inline src/connector/src/sink/big_query.rs:96]: Please update the comment.

**Review by wenym1 — APPROVED** [review]
Rest LGTM
